### PR TITLE
feat: error resilient AST to ASR

### PR DESCRIPTION
**Pull request by Pranavchiku**
Towards #5036. @certik can you please review and say if this is what the correct direction?

```console
% cat b.f90 
program assign_01
    implicit none
    integer :: x(2)
    x = []
    print *, x
end program
% lfortran b.f90            
visit_Name
semantic error: Empty array constructor is not allowed
 --> b.f90:4:9
  |
4 |     x = []
  |         ^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.

% lfortran b.f90 --continue-compilation
visit_Name
visit_Name
semantic error: Empty array constructor is not allowed
 --> b.f90:4:9
  |
4 |     x = []
  |         ^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
1    1840068032
```

```console
% cat c.f90 
program assign_01
    implicit none
    integer :: x(2)
    y = []
    print *, x
end program
% lfortran c.f90
visit_Name
semantic error: Variable 'y' is not declared
 --> c.f90:4:5
  |
4 |     y = []
  |     ^ 'y' is undeclared


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
% lfortran c.f90 --continue-compilation
visit_Name
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2669, in main
    return main_app(argc, argv);
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2612, in main_app(int, char**)
    err = compile_src_to_object_file(arg_file, tmp_o, false,
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 1000, in (anonymous namespace)::compile_src_to_object_file(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, LCompilers::CompilerOptions&, LCompilers::PassManager&)
    result = fe.get_asr2(input, lm, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 270, in LCompilers::FortranEvaluator::get_asr2(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, LCompilers::LocationManager&, LCompilers::diag::Diagnostics&)
    Result<ASR::TranslationUnit_t*> res2 = get_asr3(*ast, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 291, in LCompilers::FortranEvaluator::get_asr3(LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&)
    auto res = LFortran::ast_to_asr(al, ast, diagnostics, symbol_table,
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_to_asr.cpp", line 101, in LCompilers::LFortran::ast_to_asr(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::SymbolTable*, bool, LCompilers::CompilerOptions&)
    auto res = body_visitor(al, ast, diagnostics, unit, compiler_options,
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 3906, in LCompilers::LFortran::body_visitor(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::ASR::asr_t*, LCompilers::CompilerOptions&, std::__1::map<unsigned long long, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, LCompilers::ASR::ttype_t*, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::__1::less<unsigned long long>, std::__1::allocator<std::__1::pair<unsigned long long const, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, LCompilers::ASR::ttype_t*, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::__1::map<unsigned long long, LCompilers::ASR::symbol_t*, std::__1::less<unsigned long long>, std::__1::allocator<std::__1::pair<unsigned long long const, LCompilers::ASR::symbol_t*> > >&, std::__1::map<unsigned long long, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >, std::__1::less<unsigned long long>, std::__1::allocator<std::__1::pair<unsigned long long const, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > > >&, std::__1::map<unsigned int, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, LCompilers::ASR::ttype_t*, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::__1::less<unsigned int>, std::__1::allocator<std::__1::pair<unsigned int const, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, LCompilers::ASR::ttype_t*, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::__1::map<unsigned int, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, LCompilers::ASR::symbol_t*, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, LCompilers::ASR::symbol_t*> > >, std::__1::less<unsigned int>, std::__1::allocator<std::__1::pair<unsigned int const, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, LCompilers::ASR::symbol_t*, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, LCompilers::ASR::symbol_t*> > > > > >&, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<LCompilers::LFortran::AST::stmt_t*, std::__1::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::vector<LCompilers::LFortran::AST::stmt_t*, std::__1::allocator<LCompilers::LFortran::AST::stmt_t*> > > > >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<LCompilers::LFortran::AST::stmt_t*, std::__1::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::vector<LCompilers::LFortran::AST::stmt_t*, std::__1::allocator<LCompilers::LFortran::AST::stmt_t*> > > > > > > >&, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::vector<int, std::__1::allocator<int> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::vector<int, std::__1::allocator<int> > > > >&, std::__1::vector<LCompilers::ASR::stmt_t*, std::__1::allocator<LCompilers::ASR::stmt_t*> >&)
    b.visit_TranslationUnit(ast);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 146, in LCompilers::LFortran::BodyVisitor::visit_TranslationUnit(LCompilers::LFortran::AST::TranslationUnit_t const&)
    visit_ast(*x.m_items[i]);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4884, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::BodyVisitor>::visit_ast(LCompilers::LFortran::AST::ast_t const&)
    void visit_ast(const ast_t &b) { visit_ast_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4841, in void LCompilers::LFortran::AST::visit_ast_t<LCompilers::LFortran::BodyVisitor>(LCompilers::LFortran::AST::ast_t const&, LCompilers::LFortran::BodyVisitor&)
    case astType::mod: { v.visit_mod((const mod_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4887, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::BodyVisitor>::visit_mod(LCompilers::LFortran::AST::mod_t const&)
    void visit_mod(const mod_t &b) { visit_mod_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4474, in void LCompilers::LFortran::AST::visit_mod_t<LCompilers::LFortran::BodyVisitor>(LCompilers::LFortran::AST::mod_t const&, LCompilers::LFortran::BodyVisitor&)
    case modType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 1598, in LCompilers::LFortran::BodyVisitor::visit_Program(LCompilers::LFortran::AST::Program_t const&)
    transform_stmts(body, x.n_body, x.m_body);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 117, in LCompilers::LFortran::BodyVisitor::transform_stmts(LCompilers::Vec<LCompilers::ASR::stmt_t*>&, unsigned long, LCompilers::LFortran::AST::stmt_t**)
    this->visit_stmt(*m_body[i]);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4930, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::BodyVisitor>::visit_stmt(LCompilers::LFortran::AST::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4576, in void LCompilers::LFortran::AST::visit_stmt_t<LCompilers::LFortran::BodyVisitor>(LCompilers::LFortran::AST::stmt_t const&, LCompilers::LFortran::BodyVisitor&)
    case stmtType::Assignment: { v.visit_Assignment((const Assignment_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 2388, in LCompilers::LFortran::BodyVisitor::visit_Assignment(LCompilers::LFortran::AST::Assignment_t const&)
    this->visit_expr(*x.m_target);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4982, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::BodyVisitor>::visit_expr(LCompilers::LFortran::AST::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4649, in void LCompilers::LFortran::AST::visit_expr_t<LCompilers::LFortran::BodyVisitor>(LCompilers::LFortran::AST::expr_t const&, LCompilers::LFortran::BodyVisitor&)
    case exprType::Name: { v.visit_Name((const Name_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 8590, in LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::visit_Name(LCompilers::LFortran::AST::Name_t const&)
    visit_NameUtil(x.m_member, x.n_member, x.m_id, x.base.base.loc);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 8527, in LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::visit_NameUtil(LCompilers::LFortran::AST::struct_member_t*, unsigned long, char*, LCompilers::Location const&)
    while(ASR::is_a<ASR::StructInstanceMember_t>(*ASRUtils::EXPR(tmp_copy)) || 
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_utils.h", line 79, in LCompilers::ASRUtils::EXPR(LCompilers::ASR::asr_t const*)
    return ASR::down_cast<ASR::expr_t>(f);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr.h", line 40, in LCompilers::ASR::expr_t* LCompilers::ASR::down_cast<LCompilers::ASR::expr_t, LCompilers::ASR::asr_t>(LCompilers::ASR::asr_t const*)
    LCOMPILERS_ASSERT(f != nullptr);
AssertFailed: f != nullptr
```

### CODE REVIEW

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 1]
```diff
@@ -2405,12 +2410,21 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+                                        diag::Label("assignment here", {x.base.base.loc}),
+                                        diag::Label("declared as constant", {v->base.base.loc}, false),
+                                    }));
+                    } else {
+                        throw SemanticError(diag::Diagnostic(
+                                    "Cannot assign to a constant variable",
+                                    diag::Level::Error, diag::Stage::Semantic, {
+                                        diag::Label("assignment here", {x.base.base.loc}),
+                                        diag::Label("declared as constant", {v->base.base.loc}, false),
+                                    }));
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:2426]: I would change all `throw SemanticError` into `diag.add` + `throw SemanticAbort`, and I would only catch `SemanticAbort`.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -8438,8 +8438,16 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
                         char* x_m_id, const Location& loc) {
         if (x_n_member == 0) {
-            tmp = (ASR::asr_t*) replace_with_common_block_variables(
-                ASRUtils::EXPR(resolve_variable(loc, to_lower(x_m_id))));
+            try {
+                ASR::expr_t* expr = ASRUtils::EXPR(resolve_variable(loc, to_lower(x_m_id)));
+                tmp = (ASR::asr_t*) replace_with_common_block_variables(expr);
+            } catch (const SemanticError &e) {
+                if ( compiler_options.continue_compilation ) diag.add(e.d);
+                else throw e;
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:8446]: I would remove this part, and eventually remove `SemanticError` exception altogether.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 3]
```diff
@@ -2387,7 +2387,12 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         }
         this->visit_expr(*x.m_target);
         ASR::expr_t *target = ASRUtils::EXPR(tmp);
-        this->visit_expr(*x.m_value);
+        try {
+            this->visit_expr(*x.m_value);
+        } catch (const SemanticError &e) {
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:2392]:
```suggestion
        } catch (const SemanticAbort &e) {
```

File: src/lfortran/semantics/ast_body_visitor.cpp (hunk 1 above)
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:2426]: Something like:
```suggestion
                        diag.add(diag::Diagnostic(
                                    "Cannot assign to a constant variable",
                                    diag::Level::Error, diag::Stage::Semantic, {
                                        diag::Label("assignment here", {x.base.base.loc}),
                                        diag::Label("declared as constant", {v->base.base.loc}, false),
                                    }));
                        if (!compiler_options.continue_compilation) throw SemanticAbort();
```

**Review by certik — COMMENTED** [review]
Yes! The user behavior is exactly as it should behave, and I left some comments how to handle the internal implementation. It's very close.

**Comment by Pranavchiku** [comment]
Perfect, I’ll apply the review and make progress on this.

File: src/lfortran/semantics/ast_common_visitor.h (hunk 2 above)
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h]: I decided to keep `catch (const SemanticError &e)` and added a TODO, step by step will try to cover all cases in `tests/errors` at that time I'll replace `SemanticError` with `diag.add` + `SemanticAbort`

**Review by certik — APPROVED** [review]

**Pull request merged**